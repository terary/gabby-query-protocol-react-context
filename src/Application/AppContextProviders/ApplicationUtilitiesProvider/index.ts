import ApplicationUtilitiesProvider, {
  ApplicationUtilitiesContext,
  useTestUtilities,
} from "./ApplicationUtilitiesProvider";
import type { ApplicationUtilitiesType } from "./type";

export default ApplicationUtilitiesProvider;
export { ApplicationUtilitiesContext, useTestUtilities };
export type { ApplicationUtilitiesType };
