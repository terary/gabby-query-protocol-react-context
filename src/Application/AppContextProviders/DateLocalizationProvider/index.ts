import DateLocalizationProvider from "./Provider";

export default DateLocalizationProvider;
