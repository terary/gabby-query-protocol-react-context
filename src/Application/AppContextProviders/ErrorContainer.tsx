import { ReactNode } from "react";
import { ErrorBoundary } from "react-error-boundary";

function ErrorFallback({ error }: { error: Error }) {
  return (
    <div role="alert">
      <p>Something went wrong:</p>
      <pre>{error.message}</pre>
    </div>
  );
}

export const ErrorContainer = ({ children }: { children: ReactNode }) => {
  return <ErrorBoundary FallbackComponent={ErrorFallback}>{children}</ErrorBoundary>;
};
