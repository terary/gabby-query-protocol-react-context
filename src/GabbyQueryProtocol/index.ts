import {
  CONSTS as LIB_CONSTS,
  EXAMPLE_JSON_BLUE_SKIES as LIB_EXAMPLE_JSON_BLUE_SKIES,
} from "gabby-query-protocol-lib";

import {
  CONSTS as PROJECTION_CONSTS,
  EXAMPLE_JSON_BLUE_SKIES as PROJECTION_EXAMPLE_JSON_BLUE_SKIES,
} from "gabby-query-protocol-projection";

const GABBY_EXAMPLE_JSON_BLUE_SKIES = {
  LIB: LIB_EXAMPLE_JSON_BLUE_SKIES,
  PROJECTION: PROJECTION_EXAMPLE_JSON_BLUE_SKIES,
};

const GABBY_CONSTS = {
  LIB: LIB_CONSTS,
  PROJECTION: PROJECTION_CONSTS,
};

export { GABBY_EXAMPLE_JSON_BLUE_SKIES, GABBY_CONSTS };
