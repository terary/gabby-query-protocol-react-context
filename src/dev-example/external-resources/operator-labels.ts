// cspell:disable
// import type { TPredicateOperatorLabels } from "gabby-query-protocol-lib";
import { TPredicateOperatorLabels } from "../../lib/GabbyQueryProtocolContext";

const EN: TPredicateOperatorLabels = {
  $and: "All of these",
  $anyOf: "Any Of",
  $empty: "Is Empty",
  $eq: "Equals",
  $gt: "Greater Than",
  $gte: "Greater Than Equal",
  $isnull: "Is Null",
  $lt: "Less Than",
  $lte: "Less Than Equal",
  $like: "Contains",
  $oneOf: "One Of",
  $nand: "Not And",
  $nanyOf: "Not Any",
  $ne: "Not Equal",
  $nor: "Not Or",
  $or: "Any of these",
};
const ES: TPredicateOperatorLabels = {
  $and: "Todos estos",
  $anyOf: "Cualquiera de",
  $empty: "Is Empty",
  $eq: "Es igual a",
  $gt: "Mayor que",
  $gte: "Mayor que igual",
  $isnull: "Is Null",
  $lt: "Menor que",
  $lte: "Menor que igual",
  $like: "Contiene",
  $oneOf: "Uno de",
  $nand: "Not And",
  $nanyOf: "Not Any",
  $ne: "Not Equal",
  $nor: "Not Or",
  $or: "Cualquiera de estos",
};

const TH: TPredicateOperatorLabels = {
  $and: "ทั้งหมดนี้",
  $anyOf: "อันใดอันหนึ่ง",
  $empty: "Is Empty",
  $eq: "เท่ากับ",
  $gt: "มากกว่า",
  $gte: "มากกว่าเท่ากับ",
  $isnull: "Is Null",
  $lt: "น้อยกว่า",
  $lte: "น้อยกว่าเท่ากับ",
  $like: "มี",
  $oneOf: "หนึ่งใน",
  $nand: "Not And",
  $nanyOf: "Not Any",
  $ne: "Not Equal",
  $nor: "Not Or",
  $or: "ใดๆ เหล่านี้",
};
const AR: TPredicateOperatorLabels = {
  $and: "كل هذه",
  $anyOf: "اي من",
  $empty: "Is Empty",
  $eq: "يساوي",
  $gt: "أكبر من",
  $gte: "أكبر من يساوي",
  $isnull: "Is Null",
  $lt: "أقل من",
  $lte: "أقل من يساوي",
  $like: "يحتوي على",
  $oneOf: "واحد من",
  $nand: "Not And",
  $nanyOf: "Not Any",
  $ne: "Not Equal",
  $nor: "Not Or",

  $or: "أي منهم",
};

const SymbolsTypable: TPredicateOperatorLabels = {
  $and: "AND",
  $anyOf: "Any Of",
  $empty: "Is Empty",
  $eq: "=",
  $gt: ">",
  $gte: ">=",
  $isnull: "Is Null",
  $lt: "<",
  $lte: "<=",
  $like: "=~",
  $oneOf: "One Of",
  $nand: "Not And",
  $nanyOf: "Not Any",
  $ne: "Not Equal",
  $nor: "Not Or",

  $or: "OR",
};
const SymbolsEntities: TPredicateOperatorLabels = {
  // https://dev.w3.org/html5/html-author/charref
  $anyOf: String.fromCharCode(0x02286), // ⊆
  $empty: "Is Empty",

  $eq: "=",
  $gt: ">",
  $gte: String.fromCharCode(0x02265), // ≥
  $isnull: "Is Null",
  $lt: "<",
  $lte: String.fromCharCode(0x02264), // ≤
  $like: String.fromCharCode(0x02243), // ≃
  $oneOf: String.fromCharCode(0x02203), // ∃
  $and: String.fromCharCode(0x02227), // ∧
  $nand: "Not And",
  $nanyOf: "Not Any",
  $ne: "Not Equal",
  $nor: "Not Or",
  $or: String.fromCharCode(0x02228), // ∨
};

export { SymbolsEntities, SymbolsTypable, AR, EN, ES, TH };

// probably a more eloquent way of doing this but ran out of time.
// export const PREDICATE_OPERATORS = ["$eq", "$gt", "$lt", "$gte", "$lte", "$like", "$in"];
// export const JUNCTION_OPERATORS = ["$and", "$or"];
// export const predefinedLabels = {
//   Symbols,
//   en: Long,
//   "en-US": Long,
//   es: ES,
//   "es-MX": ES,
//   "th-TH": TH,
//   AR,
// };
