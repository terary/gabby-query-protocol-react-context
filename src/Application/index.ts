import { SUPPORTED_LOCALE_CODES_ARRAY, SUPPORTED_LOCALE_DICTIONARY } from "./i18n";
import { PageContentWrapper } from "./PageContentWrapper";
import { useApplicationUtilities } from "./AppContextProviders/hooks";
import { TSupportedLocales } from "./i18n";
import { customStyles } from "./custom-styles";
export {
  customStyles,
  PageContentWrapper,
  useApplicationUtilities,
  SUPPORTED_LOCALE_CODES_ARRAY,
  SUPPORTED_LOCALE_DICTIONARY,
};

export type { TSupportedLocales };
