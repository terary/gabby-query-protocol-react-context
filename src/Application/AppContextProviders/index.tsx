import React from "react";
import { TSupportedLocales } from "..";
import ApplicationUtilitiesProvider, {
  useTestUtilities,
} from "./ApplicationUtilitiesProvider";
import { ErrorContainer } from "./ErrorContainer";
import DateLocalizationProvider from "./DateLocalizationProvider";
import GabbyThemeProvider from "./GabbyThemeProvider";

interface IAppProviderParameters {
  children?: React.ReactNode;
  initialLocaleCode?: TSupportedLocales;
}

export const AppContextProviders = ({
  initialLocaleCode,
  children,
}: IAppProviderParameters) => {
  return (
    <ErrorContainer>
      {" "}
      {/* Error container may have issue with hooks, requires testing */}
      <ApplicationUtilitiesProvider initialLocaleCode={initialLocaleCode}>
        <GabbyThemeProvider>
          <DateLocalizationProvider>{children}</DateLocalizationProvider>
        </GabbyThemeProvider>
      </ApplicationUtilitiesProvider>
    </ErrorContainer>
  );
};

export { useTestUtilities };
