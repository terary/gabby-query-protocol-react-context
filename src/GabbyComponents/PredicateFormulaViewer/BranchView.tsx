import * as React from "react";
import Grid from "@mui/material/Grid";
import type { TPredicateProperties } from "gabby-query-protocol-lib";
// import { predicateFormulaContextUtils } from "../../GabbyQueryProtocol/PredicateFormula";
import { IconButtonRemove } from "../common/IconButtonRemove";
import { IconButtonsAdd } from "../common/IconButtonsAdd";
import { OptionSwitch } from "../common/OptionSwitch";
import { PredicateEditorPopover } from "../PredicateFormulaEditor";
import { SxProps, useTheme, Theme } from "@mui/system";
import { blue, green } from "@mui/material/colors";

import { PredicateFormulaEditorContextHooks } from "../../GabbyQueryProtocol";
const { useJunctionProperties } = PredicateFormulaEditorContextHooks;

type Props = {
  predicateId: string;
  children?: React.ReactNode;
};
export const BranchView = ({ predicateId, children }: Props) => {
  const editorContainerRef = React.useRef(null);
  const theme = useTheme();
  const [isOpenNewPredicateEditor, setIsOpenNewPredicateEditor] = React.useState(false);

  const {
    appendPredicate,
    getPredicateProperties,
    setConjunction,
    setDisjunction,
    isRoot,
    makeEmptyPredicate,
    removePredicateJunction,
  } = useJunctionProperties(predicateId);

  const bg = getPredicateProperties()?.operator === "$and" ? green[50] : blue[50];

  const handleJunctionSwitch = (operator: "$and" | "$or") => {
    if (operator === "$and") {
      setConjunction();
    } else {
      setDisjunction();
    }
  };

  const handleRemovePredicateJunction = () => {
    removePredicateJunction();
  };

  const handleAddPredicateFinishClick = (newPredicate: TPredicateProperties) => {
    // run validator?
    appendPredicate(newPredicate);
    setIsOpenNewPredicateEditor(false);
  };

  const handleAddPredicateClick = () => {
    setIsOpenNewPredicateEditor(true);
  };

  const handNewPredicateEditCancelClick = () => {
    setIsOpenNewPredicateEditor(false);
  };

  const ControlButtons = () => {
    return (
      <>
        <OptionSwitch
          operator={getPredicateProperties().operator === "$or" ? "$or" : "$and"}
          onChange={handleJunctionSwitch}
        />
        <IconButtonsAdd onClick={handleAddPredicateClick} />
        {!isRoot && <IconButtonRemove onClick={handleRemovePredicateJunction} />}
      </>
    );
  };

  const PredicateEditorWrapper = () => {
    return (
      <PredicateEditorPopover
        id={`${predicateId}-popover-editor`}
        initialPredicate={makeEmptyPredicate()}
        isOpen={isOpenNewPredicateEditor}
        onFinish={handleAddPredicateFinishClick}
        onCancel={handNewPredicateEditCancelClick}
        parentEl={editorContainerRef.current}
      />
    );
  };

  const sx = {
    // 'theme' didn't seem to work in from common utilities
    border: `1px solid ${theme.palette.text.secondary}`,
    borderRadius: "4px",
    textAlign: "center",
    backgroundColor: bg,
    height: "100%",
    m: "3px",
    p: "3px",
    width: "calc(100% - 4px)",
  } as SxProps<Theme>;

  return (
    <Grid container spacing={1} sx={sx}>
      <Grid item xs={12} ref={editorContainerRef}>
        <PredicateEditorWrapper />
        <ControlButtons />
      </Grid>
      {children}
    </Grid>
  );
};
BranchView.defaultProps = {
  children: [],
  bg: "#7FFF00",
};
