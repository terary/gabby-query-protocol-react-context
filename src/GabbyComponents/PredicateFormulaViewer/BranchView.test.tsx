import React from "react";
import renderer from "react-test-renderer";
import {
  render,
  screen,
  act,
  fireEvent,
  cleanup,
  within,
  getByTestId,
  getByLabelText,
  findByDisplayValue,
  getByText,
  findByText,
  RenderResult,
} from "@testing-library/react";
import "@testing-library/jest-dom";
import userEvent from "@testing-library/user-event";
import selectEvent from "react-select-event";
import { AppContextProviders } from "../../Application/AppContextProviders";

import {
  PredicateFormulaEditorFactory,
  TPredicateProperties,
  TPredicatePropertiesArrayValue,
  TValidatorResponse,
} from "gabby-query-protocol-lib";

//import { AppContextProviders } from "../Application/AppContextProviders";
// import GQPPredicateEditorContextProvider from "../../GabbyQueryProtocol/PredicateFormula/context";
import { PredicateFormulaEditorContext } from "../../GabbyQueryProtocol";
//import GQPPredicateEditorContextProvider from "../../GabbyQueryProtocol/contexts/GQPPredicateFormula/context";
import { GABBY_EXAMPLE_JSON_BLUE_SKIES } from "../../GabbyQueryProtocol";
import * as predefinedOperatorLabels from "../../GabbyQueryProtocol/external-resources/operator-labels";

import { BranchView } from "./BranchView";

const subjectDictionaryJson =
  GABBY_EXAMPLE_JSON_BLUE_SKIES.LIB.predicateSubjectsDictionaryJson;
const predicateFormula = GABBY_EXAMPLE_JSON_BLUE_SKIES.LIB.predicateTreeJson;

const predicateFormulaEditor = PredicateFormulaEditorFactory.fromJson({
  subjectDictionaryJson,
  predicateTreeJson: predicateFormula,
});

type PredicateTypes = TPredicateProperties | TPredicatePropertiesArrayValue;

const validator = (predicateProperties: PredicateTypes) => {
  return { hasError: false, errorMessages: [] };
};

const noopFinish = (predicate: PredicateTypes) => {};
const noopCancel = () => {};
type Props = {
  predicateId: string;
  children?: React.ReactNode;
};

const PredicateEditorWithApplicationContext = ({ predicateId, children }: Props) => {
  return (
    <AppContextProviders>
      <PredicateFormulaEditorContext.Provider
        predicateFormulaEditor={predicateFormulaEditor}
        operatorLabels={predefinedOperatorLabels.EN}
      >
        <BranchView predicateId={predicateId} />
        {/* <PredicateEditor
        initialPredicate={predicateFormulaEditor.subjectDictionary.makeEmptyPredicate()}
        onFinish={(p: any) => {}}
        onCancel={() => {}}
        testSpyPredicateOnChange={testSpyPredicateOnChange}
        validator={validator}
      />{" "} */}
      </PredicateFormulaEditorContext.Provider>
    </AppContextProviders>
  );
};
describe("BranchView - default snapshot", () => {
  afterEach(cleanup);
  // describe("SubjectChange", () => {
  //   it.only("Should change subject, reset operator and empty value", async () => {
  //     const changeObserver = jest.fn((p: any) => {
  //       console.log(p);
  //       renderer;
  //     });

  //     const mockFinish = jest.fn();

  //     const result = render(
  //       <PredicateEditorWithApplicationContext
  //         testSpyPredicateOnChange={changeObserver}
  //         onFinish={mockFinish}
  //       />
  //     );
  //   });
  // });
  describe("Snapshots", () => {
    it("renders correctly", () => {
      const tree = renderer
        .create(
          <PredicateEditorWithApplicationContext
            predicateId={predicateFormulaEditor.rootNodeId}
          />
        )
        .toJSON();
      expect(tree).toMatchSnapshot();
    });
  });
});
