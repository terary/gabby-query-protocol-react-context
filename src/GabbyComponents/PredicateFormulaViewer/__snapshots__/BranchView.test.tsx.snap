// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BranchView - default snapshot Snapshots renders correctly 1`] = `
Array [
  " ",
  <div
    className="Toastify"
  />,
  <div
    className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-1 css-1rhdp4s-MuiGrid-root"
  >
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1em9ut6-MuiGrid-root"
    >
      <label
        className="MuiFormControlLabel-root MuiFormControlLabel-labelPlacementStart css-1f2kvjf-MuiFormControlLabel-root"
      >
        <span
          className="MuiSwitch-root MuiSwitch-sizeMedium css-julti5-MuiSwitch-root"
        >
          <span
            className="MuiSwitch-switchBase MuiSwitch-colorPrimary MuiButtonBase-root MuiSwitch-switchBase MuiSwitch-colorPrimary PrivateSwitchBase-root css-5ryogn-MuiButtonBase-root-MuiSwitch-switchBase"
            onBlur={[Function]}
            onContextMenu={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            style={
              Object {
                "color": "#e8f5e9",
              }
            }
            tabIndex={null}
          >
            <input
              checked={false}
              className="MuiSwitch-input PrivateSwitchBase-input css-1m9pwf3"
              onChange={[Function]}
              type="checkbox"
              value="top"
            />
            <span
              className="MuiSwitch-thumb css-jsexje-MuiSwitch-thumb"
            />
          </span>
          <span
            className="MuiSwitch-track css-1yjjitx-MuiSwitch-track"
          />
        </span>
        <span
          className="MuiTypography-root MuiTypography-body1 MuiFormControlLabel-label css-ahj2mt-MuiTypography-root"
        >
          Any of these
        </span>
      </label>
      <button
        className="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorSuccess MuiIconButton-sizeSmall css-15ghhi9-MuiButtonBase-root-MuiIconButton-root"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onContextMenu={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-ptiqhd-MuiSvgIcon-root"
          data-testid="AddCircleOutlineIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M13 7h-2v4H7v2h4v4h2v-4h4v-2h-4V7zm-1-5C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.41 0-8-3.59-8-8s3.59-8 8-8 8 3.59 8 8-3.59 8-8 8z"
          />
        </svg>
      </button>
    </div>
  </div>,
]
`;
